File: projecteuler/go/15/15.go
/*

Lattice paths

Starting in the top left corner of a 22 grid, there are 6 routes
(without backtracking) to the bottom right corner.

How many routes are there through a 2020 grid?

*/
package main

import (
	"fmt"
	"github.com/zeroshirts/challenges/projecteuler/go/misc"
	"math/big"
)

// square grid
// no backtracking allowed
func numRoutes(size int64) int64 {
	// note: r1,r2,d1,d2 is same route as r2,r1,d2,d1

	n := size + size
	k := size

	nfactorial := misc.Factorial(n)
	kfactorial := misc.Factorial(k)
	nMinuskfactorial := misc.Factorial(n - k)

	C := new(big.Int).Div(nfactorial, new(big.Int).Mul(kfactorial, nMinuskfactorial))
	return C.Int64()
}

func main() {
	fmt.Println(numRoutes(20))
}

File: projecteuler/julia/15/15.jl
#=

Lattice paths

Starting in the top left corner of a 2x2 grid, there are 6 routes
(without backtracking) to the bottom right corner.

(note: see image p015.gif)

How many routes are there through a 20x20 grid?

=#

using Test

# square grid, no backtracking
function numRoutes(gridSize)
  #=
  n = gridSize + gridSize
  n! permutations
  
  Overcounting by gridSize! * gridSize!
  e.g., r1 r2 d1 d2 === r2 r1 d2 d1
  and r1 r2 d1 d2 === r2 r1 d1 d2
  and r1 r2 d1 d2 === r1 r2 d2 d1
  
  =#
  n = big(gridSize + gridSize)
  p = factorial(n)
  return div(div(p, factorial(gridSize)), factorial(gridSize))
end

@time @test numRoutes(2) == 6
@time @test numRoutes(20) == 137846528820
